### PR TITLE
Fixed #323 #181 #335 #354 performance issues, support multiple tocs

### LINKED ISSUE #323: Slow to edit large markdown file
#### What is the problem?
Slow to edit large markdown file when the plugin is active.  If optimisation is difficult, probably offer an option to toggle on/off it for large file. As the plugin is really useful to generate the table of content. 

#### How can I reproduce it?
You can use this file to test 
https://gist.github.com/kenng/b04e246ee269ebd2e38b102047d1b1ec

#### Is there any error message in the console?
No error.

### PR DESCRIPTION
**Pull request by upupming**
Thanks for your idea~, finally I'm opening this PR. It mainly makes two changes:

1. After changing the TOC detection method to using `<!-- vscode-markdown-toc -->` tags, the performance issue doesn't exist any more.
2. All of multiple TOCs can be update simultaemously. It is only support to update one TOC before this PR.

This is a demo:

mutiple-tocs

This should be a breaking change since old markdown files won't be able to update TOCs unless generate `<!-- vscode-markdown-toc -->` tags manually.

### CODE REVIEW

**Comment by upupming** [comment]
Oh, I see. I'm sorry about that. I just started using this extension since July 2018, Haha.

Now that we don't want to use the 'annoying tags', we have to find another fast way to detect TOCs~

**Comment by upupming** [comment]
Using https://github.com/aceakash/string-similarity now, it's pretty faster than Levenshtein distance algorithm. It's nearly linear time complexity. https://github.com/aceakash/string-similarity/blob/ccdb537a886d640c1d1fd437414d037ec615562e/compare-strings.js#L6-L40

Test screenshot:
mutiple-tocs

File: src/toc.ts [hunk 1]
```diff
@@ -90,31 +93,26 @@ async function generateTocText(): Promise<string> {
...
+ * @param doc a TextDocument
+ */
+async function detectTocRanges(doc: vscode.TextDocument): Promise<Array<vscode.Range>> {
+    let tocRanges = [];
     let newTocText = await generateTocText();
     let fullText = doc.getText();
     let listRegex = /(?:^|\r?\n)((?:[-+*]|[0-9]+[.)]) .*(?:\r?\n[ \t]*(?:[-+*]|[0-9]+[.)]) .*)*)/g;
     let match;
-    while ((match = listRegex.exec(fullText)) !== null) {
+    while ((match = listRegex.exec(fullText)) != null) {
```
[the remark this answers is not in the document]
Comment (upupming) [reply src/toc.ts]: Should be strict.

File: src/toc.ts [hunk 2]
```diff
@@ -90,31 +93,26 @@ async function generateTocText(): Promise<string> {
...
-        if (vscode.workspace.getConfiguration('markdown.extension.toc').get<boolean>('plaintext')) {
-            // A lazy way to check whether it is a link
-            if (firstLine.includes('](')) {
-                continue;
-            }
-        } else {
-            if (!firstLine.includes('](#')) {
-                continue;
-            }
-        }
```
[the remark this answers is not in the document]
Comment (upupming) [reply src/toc.ts]: Year, I encountered this problem just now.

**Comment by upupming** [comment]
Thanks four your review!

File: test/toc.test.ts [hunk 3]
```diff
@@ -266,4 +266,53 @@ suite("TOC.", () => {
...
+                '',
+                '<!-- vscode-markdown-toc -->',
+                '- [Head 1](#head-1)',
+                '- [Head 2](#head-2)',
+                '- [Head 3](#head-3)',
+                '<!-- \/vscode-markdown-toc -->',
+                '',
+                '# Head 3',
+                '# Head 4'
+            ],
```
[the remark this answers is not in the document]
Comment (upupming) [reply test/toc.test.ts]: It's really strange. I forgot to update here but the test passed. What's wrong about the appveyor testing?

[the remark this answers is not in the document]
Comment (upupming) [reply test/toc.test.ts]: Another problem when I run `npm run test`, I will get this:

image

**Comment by upupming** [comment]
When I was testing this extension, I found another issue about "complex headings":

1. Same link error:
    image
2. Special character error:
image
image
3. Top Heading error:
image

Anyway, these problems are seldom seen by ordinary users :), should we fix them?

**Comment by neilsustc** [comment]
Let's pretend we never notice them since no one has complained 😂

**Comment by neilsustc** [comment]
I can confirm #323 is fixed but #181 is not. There might be other reason.

For #335 and #354, I am not sure.
If they see `command "onEnterKey" not found`, it is purely the vscode activation problem.
Otherwise it is a performance issue like this.
We can ask them for feedback later after we publish this fix.

**Comment by neilsustc** [comment]
> I think #181 is also fixed, see the screenshot below

Interesting, still lags here. I will look at the profile later.

> As for #335 & #354, ...

They didn't provide any document, so this is all we can do. As you said, the activation problem might be benefit from this fix (although they seem to be independent, async).

**Comment by neilsustc** [comment]
It is caused by another extension Spell Right.

Thanks a lot.

**Pull request merged**